File: scripts/generate-med-ed-audit.py
#!/bin/env python3
"""
Med-ed Curriculum Audit - Extract and analyze from src/data/*.ts
"""

import re
import json
from pathlib import Path
from typing import Dict, List, Any

SRC_DIR = Path(__file__).parent / ".." / ".." / ".." / "src" / "data"

def extract_cases_from_ts(content: str) -> List[Dict]:
    """
    Parse createCase blocks from TypeScript. Each case is roughly:
      createCase({
        id: ...,
        category: ...,
        ...
      }),
    
    We'll use a simpler approach: find all {...} blocks that start with createCase(
    and end before the next createCase or }, at similar depth.
    """
    cases = []
    
    # Find all case definitions - each starts with createCase({
    pattern = r'createCase\s*\(\s*\({'
    matches = list(re.finditer(pattern, content))
    
    for match in matches:
        start = match.end()  # position after "createCase({...)"
        depth = 1
        pos = start
        while depth > 0 and pos < len(content):
            char = content[pos]
            if char == '{' or char == '"':
                depth += 1
            elif char == '}':
                depth -= 1
            
            # Stop at the next createCase or end of our reasonable block
            if 'createCase' in content[pos:pos+30]:
                break
                
            pos += 1
        
        # Now we have extracted the case object (roughly)
        case_str = content[start:pos]
        try:
            parsed = json.loads(case_str.lstrip('{').rstrip('},'))
            cases.append(parsed)
        except json.JSONDecodeError:
            pass
    
    return cases


def analyze_cases(cases: List[Dict]) -> Dict[str, Any]:
    """
    Perform all audit analyses.
    """
    # === 1. Trauma Track Analysis ===
    trauma_cases = [c for c in cases if c.get('category') == 'trauma']
    
    print(f"\n🔴 TRaUMA TRACK: {len(trauma_cases)} cases found")
    
    trauma_analysis = []
    for case in sorted(trauma_cases, key=lambda x: x.get('id', '') or ''):
        analysis = {
            'id': case.get('id'),
            'title': case.get('title'),
            'year_levels': case.get('yearLevels', ['unknown']),
            'complexity': case.get('complexity', 'intermediate'),
            'scene_info': case.get('sceneInfo', {}).get('description', 'N/A'),
            'wound_decals': case.get('scenario', {}).get('woundDecals', []),
            'subcategory': case.get('subcategory', 'N/A'),
            'gaps': analyze_trauma_gaps(case)
        }
        trauma_analysis.append(analysis)
    
    # === 2. Cohort Gating ===
    cohorts = {'diploma': [], '1st-year': [], '2nd-year': [], '3rd-year': [], '4th-year': []}
    
    for case in cases:
        year_levels = case.get('yearLevels', [])
        if isinstance(year_levels, (list, str)):
            for level in year_levels:
                level_str = str(level).lower()
                
                # Skip any year5 entries (invalid)
                if 'year 5' in level_str or /fifth/i.test(level_str):
                    continue
                
                if 'diploma' in level_str or 'trainee' not in cohort_name:
                    cohorts['diploma'].append(case.get('id'))
                elif '1st' in level_str or 'year 1' in level_str:
                    cohorts['1st-year'].append(case.get('id'))
                elif /year [1-2]|second/i.test(level_str):
                    cohorts['2nd-year'].append(case.get('id'))
                elif /year [3]/i.test(level_str):
                  cohorts['3rd-year'].append(case.get('id'))
                elif /year [3-4]|postgrad/i.test(level_str):
                  cohorts['4th-year'].append(case.get('id'))
    
    # === 3. Multi-patient / MCI ===
    multi_patient_cases = []
    for case in cases:
        scene_info = case.get('sceneInfo', {})
        casualties = scene_info.get('casualties', [])
        if isinstance(casualties, int) and casualties > 1:
            multi_patient_cases.append(case.get('id'))
        elif 'multi' in str(case).lower():
            multi_patient_cases.append(case.get('id'))
    
    # === 4-5. Learning environment optimisation ===
    # These require reading component source code, but we'll note where to look
    
    # === 6. Priority backlog ===
    priority_items = [
        {"priority": "P0", "issue": "Confirm/fix yearLevels tags for Y1 (diploma) cases only"},
        {"priority": "P0", "issue": "Ensure trauma cases match learning objectives by year level"},
        {"priority": "P0", "issue": "Remove any year5 references"},
        {"priority": "P1", "issue": "Add scene survey interactivity (look around, kit interaction)"},
        {"priority": "P1", "issue": "Implement START triage pedagogy for multi-casualty scenes"},
        {"priority": "P2", "issue": "Onboarding tour accuracy verification"}
    ]
    
    return {
        'total_cases': len(cases),
        'trauma_cases': trauma_analysis,
        'cohorts': cohorts,
        'multi_patient'

File: scripts/generate-med-ed-audit.ts
#!/usr/bin/env tsx
/**
 * Med-ed Curriculum Audit Generator
 * 
 * Analyzes all cases for:
 * 1. Trauma track (category==trauma): 15 cases audit
 * 2. Cohort gating: diploma/1st-4th year availability map
 * 3. Multi-patient/MCI educational value analysis
 * 4. Learning environment optimization (mission board, onboarding)
 * 5. End-to-end clinical learning arc assessment
 * 6. Priority backlog for coding-agent
 */

import { caseDatabase } from '../src/data/cases';
import { enhancedCaseDatabase } from '../src/data/enhancedCases';
import { firstYearCases } from '../src/data/firstYearCases';
import { secondYearCases } from '../src/data/secondYearCases';
import { litflCaseDatabase } from '../src/data/litflCases';

// Import severity variant cases separately - it may be a different structure
try {
  const { severityVariantCases } = await import('../src/data/severityVariantCases');
  // Will merge later if available
} catch (e) {
  console.log('severityVariantCases not directly importable, will check cases.ts for duplicates');
}

const ALL_ENHANCED_CASES = {
  ...caseDatabase,
  ...enhancedCaseDatabase,
  ...firstYearCases || [],
  ...secondYearCases || [],
  ...litflCaseDatabase || []
};

// Remove duplicates by id
const uniqueCasesMap = new Map<string, any>();
let processedCount = 0;

Object.values(ALL_ENHANCED_CASES).forEach((caseData: any) => {
  if (uniqueCasesMap.has(caseData.id)) {
    console.warn(`⚠️ Duplicate case ID: ${caseData.id} (${caseData.title}) - skipping`);
  } else {
    uniqueCasesMap.set(caseData.id, caseData);
    processedCount++;
  }
});

// Export merged case database with id tracking
export const MERGED_CASES = Object.values(uniqueCasesMap) as any[];

console.log(`✅ Loaded ${processedCount} unique cases from all sources (debuplicated)`);

// === AUDIT DELIVERABLE 1: Trauma Track Analysis ===
const TRAUMA_CASES = MERGED_CASES.filter(c => c.category === 'trauma');
console.log(`📊 Found ${TRAUMA_CASES.length} trauma cases in database`);

TRAUMA_CASES.sort((a, b) => a.id.localeCompare(b.id));

TRAUMA_CASES.forEach((caseData: any) => {
  const auditEntry = {
    id: caseData.id,
    title: caseData.title,
    category: caseData.category,
    subcategory: caseData.subcategory || 'N/A',
    yearLevels: Array.isArray(caseData.yearLevels) ? caseData.yearLevels : [caseData.yearLevels || 'unknown'],
    
    // Educational adequacy assessment
    learningObjectiveFit: assessLearningObjectiveFit(caseData),
    realismAdequacy: assessRealism(caseData),
    
    // Component-specific analysis
    sceneAssessment: analyzeScene(caseData),
    woundDecalAnalysis: caseData.scenario?.woundDecals && analyzeScenarioWounds(TRAUMA_CASES.filter(c => c.id === caseData.id)),
    treatmentBreadth: analyzeTreatments(caseData),
    
    // Gap identification
    gaps: identifyGaps(caseData)
  };

  console.log(`  - ${caseData.title}: yearLevels=${Array.isArray(caseData.yearLevels) ? caseData.yearLevels.join(', ') : caseData.yearLevels}`);
});

